### ISSUE TITLE
Reading : One Network, Many Masks. Towards More Parameter-Efficient Transfer Learning

### ISSUE THREAD

**Issue by tikiti-kyotaro** [tikiti-kyotaro/Paperlog#19]
conference : ACL 2023
URL : https://arxiv.org/abs/2305.17682

**Abstract**
Fine-tuning pre-trained language models for multiple tasks tends to be expensive in terms of storage. To mitigate this, parameter-efficient transfer learning (PETL) methods have been proposed to address this issue, but they still require a significant number of parameters and storage when being applied to broader ranges of tasks. To achieve even greater storage reduction, we propose PROPETL, a novel method that enables efficient sharing of a single PETL module which we call prototype network (e.g., adapter, LoRA, and prefix-tuning) across layers and tasks. We then learn binary masks to select different sub-networks from the shared prototype network and apply them as PETL modules into different layers. We find that the binary masks can determine crucial information from the network, which is often ignored in previous studies. Our work can also be seen as a type of pruning method, where we find that overparameterization also exists in the seemingly small PETL modules. We evaluate PROPETL on various downstream tasks and show that it can outperform other PETL methods with approximately 10% of the parameter storage required by the latter.1

# 1. どんなもの？
PETLはパラメータ量などを減らすために考えられたものの、幅広いタスクに応用するにはまだストレージ量などに問題あり
→タスク間などでLoRAなどを簡単に共有するPROPETLを提案
（プロトタイプネットワークとはLoRA, AT, prefix-tuningなどを示す）

PROPETLはバイナリマスクを学習してプロトタイプネットワークのサブネットワークを選択
→違う層にそれを再利用
→バイナリマスクはネットワークの中から普段無視されるような重要な情報を選択可能であることがわかった

モデルのアビリティはパラメータ数だけでなく、サブネットワークの構造も重要なファクターである
→PETLに対してサブネットワークの構造も考慮させたい
![[スクリーンショット 2023-06-17 16.15.38.png]]
PROPETLの図
PROPETLはATなどのプロトタイプネットワークを異なる層やタスクで共有することを可能にする
→Maskを使うことでサブネットワークを選択？

ある層で使われたプロトタイプネットワークは違うマスクを使えば違う層に適応可能

貢献
1. 構造も考慮しながらプルーニングするPROPETLを提案
2. パラメータ効率を上げながら性能も向上
3. バイナリマスクを使うという新たなプルーニングの考え方を提案

# 2. 先行研究に比べてどこがすごい？
今まではただパラメータを減らすためにプルーニングしていたが、今回の手法は構造も考慮する

# 3. PROPETL
![[スクリーンショット 2023-06-17 16.25.35.png]]
マスクの利用（左）
1. プロトタイプネットワーク（灰色を生成）
2. 各層におけるマスクを生成

プロトタイプネットワークのパラメータ数をnとすると層の数だけパラメータ数を分割できる？

最終的な目的関数
![[スクリーンショット 2023-06-17 17.11.27.png]]
θ_pro : プロトタイプのネットワーク
m_i, i番目の層でのネットワークのマスク
Y_i : i番目の入力の正解ラベル
X_i : i番目の入力のセンテンス
θ_lm : 言語モデルのパラメータ（固定）

マスクの生成
1. 各バイナリマスクに対してスコア sl を設定
2. 上位k%のslの絶対値を持つ部分を１に設定、それ以外を０にする（関数h）
3. 誤差逆伝播によって最適化

![[スクリーンショット 2023-06-17 17.26.44.png]]
PROPETLの疑似コード
1. プロトタイプネットワークの初期化
2. マスクのスコアsの初期化
3. 訓練データ内を走査
	1. 各層を走査
		1. マスクを閾値関数hを用いて生成
		2. θ_sub（θ_proとマスクの内積）を生成
	2. θ_proを誤差逆伝播で最適化
	3. sを誤差逆伝播で最適化
4. 各層を走査
	1. マスクを生成
5. θ_proとマスクを返す

# 4. 実験設定
データセット：GLUE, Xsum, WMT Ro-EN
モデル：RoBERTa（GLUE）, T5（生成とマルチタスク）

# 5. 実験結果
## 5.1 NLU
![[スクリーンショット 2023-06-17 17.47.17.png]]
PROPETLが元の手法を上回る？
パラメータはマスク+ネットワークの分少し増加
容量は1/9
PROPELT_adapterは0.1%の容量でFTとほぼ同じ性能
PROPETL_prefix-tuningはあんまりうまくいかない
→prefix-tuningの再パラメータ化がマスク学習に悪影響を与えるためと考えられる

## 5.2 生成タスク
![[スクリーンショット 2023-06-17 18.01.07.png]]
(a), (b)が生成タスクの結果
adapterがメモリの効率よく高い性能を出せる
XSum は3.2%の容量でFTを上回る

Ro-Enはうまくいかない
→翻訳タスクは難しく、より多くのパラメータが必要となるから
→PETLでは難しい生成タスクはうまくいかないことが知られている！！！！！！

(c)は分類タスク
0.002%のメモリ量でFTの性能を上回った

## 5.3 マルチタスク
![[スクリーンショット 2023-06-17 18.06.18.png]]
PROPETL_adapter のbn=64が一番高い性能
タスク間転移に特化して開発されたHyperformer++との比較
→パラメータは約26倍であるものの正解率は0.5%ほどしか向上していない

vanilla AT との比較
→パラメータ数が1/9になっているにも関わらず、性能が向上している

ボトルネックのサイズが6の場合の実験
→通常のATは85.48から83.94に低下
→PROPETLを用いれば、bn=6でもFTと同等の性能を保っている
→通常のATはボトルネックを小さくしてしまうとよくない
→ボトルネックを小さくしてもうまくいくのはmaskを用いることでサブネットワークの構成も学習しているから？

![[スクリーンショット 2023-06-19 15.35.03.png]]
PROPETLは大規模モデルに対してはより効果的であることを示した例（少なすぎない...？）

# 6. 議論はあるか？
## 6.1 PROPETLとATのスケールの違い
1. PROPETL_adapterとATはスケールを増やすにつれて同じような性能の増加が見られる
2. 容量を増やしていくと性能が収束していく
3. 大幅にFTを超える性能を出すことはできない
4. PROPETLの方が早くFTと同じくらいの性能を出すことができる
5. 同じ容量ならPROPETLの方が性能が高い

## 6.2 共有とマスキングどっちがいい？
マスクだけするか共有だけするかで比較
![[スクリーンショット 2023-06-19 18.20.17.png]]

1. マスクだけを学習する
	パラメータ数が大幅に増加
	マスクの方がいい結果を出す
2. マスクの学習せずに共有だけする

マスキング：構造情報をサブネットワークを注入
シェアリング：パラメータサイズを大幅に削減
両方やることが大事

## 6.3 サブネットワークのサイズによる影響
サブネットワークのハイパラkを変更して実験
![[スクリーンショット 2023-06-19 18.26.05.png]]
0.1~0.5までは性能が向上
0.5~1.0までは性能が低下

0.5%が最高！！
増やしすぎるとモデルの重なりが生まれてしまうため、構造の情報が埋もれてしまう
構造情報とサブネットワークで使用されるパラメータ数のバランスを調整する必要がある

## 6.4 PROPETLとPETLの関係性
